### PR TITLE
refactor caching options and require own cache instance

### PR DESCRIPTION
**Pull request by doits**
To enable caching, you must pass a cache store instance:

```rb
class UserSerializer
  include FastJsonApi::ObjectSerializer

  # enable caching
  cache_store ActiveSupport::Cache::MemoryStore.new
end
```

To disable it, simply pass a falsy value:

```rb
class AdminUserSerializer < UserSerializer
  # disable caching
  cache_store false
end
```

Set race condition ttl and namespace:

```rb
class UserSerializer
  include FastJsonApi::ObjectSerializer

  # setting all cache options is now possible
  cache_store ActiveSupport::Cache::MemoryStore.new(
    race_condition_ttl: 20.seconds, namespace: 'my_caching_namespace'
  )
end
```

- Very explicit where cache is stored
- No `Rails.cache` automagic with possible colliding keys
- Rails 5.2+ cache versioning support by using a cache that supports it
- Implicit namespace support (by cache instance, e.g.
  `ActiveSupport::Cache::MemoryStore.new(namespace: 'jast_jsonapi')`
- All cache options can be set directly, no need to for fast jsonapi to handle them

refs #51

## Checklist

Please make sure the following requirements are complete:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes /
  features)
- [x] All automated checks pass (CI/CD)

### CODE REVIEW

**Comment by stas** [comment]
@doits looks good so far!

Minor feedback:
 * [ ] move `cache_store` into `cache_options(store: ...)`, there's no need to introduce a new API method
 * [ ] remove `cached` or `cache_options(enabled: true)`, if the cache store is present, it's clear that it's enabled
 * [ ] add deprecated API warnings for `enabled: true` and for missing `store: <CACHE INSTANCE>` which is now being required

Thanks for working on this!

**Comment by doits** [comment]
> move cache_store into cache_options(store: ...), there's no need to introduce a new API method

Alright, changed it, thought there's no need for passing a hash because there is no other cache option. But maybe there will be in the future, so let's keep it.

>  remove cached or cache_options(enabled: true), if the cache store is present, it's clear that it's enabled

:+1:

> add deprecated API warnings for enabled: true and for missing store: <CACHE INSTANCE> which is now being required

I added a hard error in this case. It *could* fall back to `Rails.cache` when *only* `enabled: true` is passed, but `cache_length` and `race_condition_ttl` have no effect anymore unless a deprecated call to `.fetch` is used when serializing the record. Which might slow down the serializer (another if/else when serializing the record).

**Comment by stas** [comment]
> Alright, changed it, thought there's no need for passing a hash because there is no other cache option. But maybe there will be in the future, so let's keep it.

I don't follow. If we don't allow extra parameters, how are we going to pass information about `expires_in: , race_condition_ttl: `?! The plan was to introduce the `store: ` to be explicit about what cache is being used and any cache-specific options to be removed from our concern. Remove the `fetch()` specific parameters will downgrade the functionality we already provide.



> I added a hard error in this case. It _could_ fall back to `Rails.cache` when _only_ `enabled: true` is passed, but `cache_length` and `race_condition_ttl` have no effect anymore unless a deprecated call to `.fetch` is used when serializing the record. Which might slow down the serializer (another if/else when serializing the record).

We can not introduce a hard-error. The idea is to not introduce a breaking change, just warn about how the functionality changed. So if `store:` is not passed, we will try to use `Rails.cache`, and warn the users about it. I also don't understand why the other parameters have no longer effect? Any other options passed as part of the `cache_options()`, except for `store:` will be used for `fetch()`, I don't see how else this is going to work...

@doits please take another look. Maybe I'm missing something, but I think we're removing things that are expected to be present.

**Comment by doits** [comment]
> I don't follow. If we don't allow extra parameters, how are we going to pass information about expires_in: , race_condition_ttl: ?!

You pass parameters directly to your store:

```rb
cache_options store: ActiveSupport::Cache::MemoryStore.new(
  race_condition_ttl: 20.seconds,
  namespace: 'my_caching_namespace',
  expires_in: 100.minutes,
  whatever_option_the_cache_supports: :is_possible
)
```

> Remove the fetch() specific parameters will downgrade the functionality we already provide.

Not if you follow the example above. Now *all* options are possible, not only arbitrary hard coded ones. And you define them all on your store, all at the same place.

In theory it could still be like this:

```rb
cache_options \
  store: ActiveSupport::Cache::MemoryStore.new(
    namespace: 'my_caching_namespace',
    whatever_option_the_cache_supports: :is_possible
  ),
  race_condition_ttl: 20.seconds,
  expires_in: 100.minutes
```

... but this feels wonky that you define some or all options directly on your instance and just two others in the fast-jsonapi options.

> We can not introduce a hard-error. The idea is to not introduce a breaking change, just warn about how the functionality changed. So if store: is not passed, we will try to use Rails.cache, and warn the users about it.

I thought about this, but then you must have a thing like this when serializing a record (pseudo-code):

```rb
if old_cache_params_were_passed
  Rails.cache.fetch(
    record.cache_key,
    race_condition_ttl: cache_options[:race_condition_ttl],
    expires_in: cache_options[:expires_in]
  )
else
  cache_instance.fetch(record)
end
```

I have no benchmark but I thought avoiding an additional if/else while serializing a record might be worthwhile. Maybe a simple check for a boolean has no impact, then it only makes the logic a little bit more complicated.

But if staying backwards compatible 100 % is the long term goal, then something like this must be done.

>  I also don't understand why the other parameters have no longer effect? Any other options passed as part of the cache_options(), except for store: will be used for fetch(), I don't see how else this is going to work...

The options were only needed for the `.fetch` because the `Rails.cache` was used by default, so there was no other way to do this. But if you define your own cache store - which you always should - there is no need for this.

**Comment by stas** [comment]
@doits the `fetch()` accepts and uses all those parameters, so they should stay. Generally speaking, options like `expire_in:` are used per-call many times, so assuming that every `store:` will be configured in isolation for each serializer is not fair.

As for defaulting to the `Rails.cache`, I'm not sure I understand what performance penalty you're describing since `store:` is checked and set-up just once, when we call `cache_options()`. Consider:
```
def cache_store(opts)
  if opts[:store].nil?
    warn('DEPRECATION WARNING: `store:` is required, we will default to `Rails.cache` if available.')
    self.cache_store_instance = Rails.cache if defined?(Rails)
  else
    self.cache_store_instance = opts[:store]
  end
  ...
end
```

**Comment by doits** [comment]
> the fetch() accepts and uses all those parameters, so they should stay. Generally speaking, options like expire_in: are used per-call many times, so assuming that every store: will be configured in isolation for each serializer is not fair.

That is true. But I thought one of the reasons to refactor it - besides the obvious point to be able to specify an own cache store - was to support more cache options, for example [all those](https://github.com/rails/rails/blob/80e72c5eb7d4f79cc8eaa63237e5a60dff48f4f4/activesupport/lib/active_support/cache.rb#L22-L24) and whatever else the specific cache store supports.

If only those two params `expires_in` and `race_condition_ttl` are kept for `.fetch`, I don't understand why others are not. Users will not understand where to configure what option for their cache either.

And doesn't explicitly having them contradict with the "less rails-specific API calls" checkbox of #51, because all these are Rails' cache options? (Even though currently `cache_length` gets renamed to `expires_in`, which only makes sense for a Rails' cache and additionally makes reasoning about it more complicated.)

If it should be configurable per serializer without the need to pass a store - which I see a point to when a store instance was inherited -, maybe then the cache options should be saved as a hash and applied generally?

```rb
def cache_options(opts)
  opts = opts.dup # do not change passed in hash

  if opts[:store].nil? && !cache_store_instance # store might exist already by inheritance
    warn('DEPRECATION WARNING: `store:` is required, we will default to `Rails.cache`.')
    self.cache_store_instance = Rails.cache

    # for backwards compatibility set and rename old options
    opts[:race_condition_ttl] ||= 5.seconds
    opts[:expires_in] ||= opts.delete(:cache_length) || 5.minutes
    opts.delete(:enabled)
  else
    self.cache_store_instance = opts.delete(:store)
  end

  # to make sure existing options are not overwritten, merge with old options
  self.cache_store_options = (cache_store_options || {}).merge(opts)
end

# ... later in the serializer ...

cache_store_instance.fetch(record, *cache_store_options)
```

(I'd also let it hard fail if `Rails.cache` is not available, because otherwise a user of a non Rails project might think caching was enabled and it will silently not do any caching at all.)

With the above we have:

- [x] 99 % backwards compatibility
  - [x] Old options still work when *only* using old syntax, you cannot mix old and new
  - [x] Old defaults are set *only* when using old options
  - [ ] Passing `cache_options enabled: false` disables the cache
- [x] Own cache store instance can be supplied
- [x] All cache options are configurable at the same place
- [x] There are no arbitrary defaults set when using new options (e.g no default `race_condition_ttl`)
- [x] Specific cache options can be overridden per serializer
- [x] All cache options are inherited

Still not 100 % compatible to the old API, but maybe 99 % are enough? I mean you could still add a check for `cache_options enabled: false` and disable all caching then, but I'm not sure if this is really the way to go only to stay 100 % compatible ...

**Comment by stas** [comment]
@doits could you please update the PR, it's not easy to review code snippets.

Also we should drop the default values for cache options, consider the case were the provided `store:` doesn't support them when calling `fetch()`. Not to mention that it's not this library concern to configure those values. [ActiveSupport has already safe defaults for those values](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/cache/mem_cache_store.rb#L156) so we should be good and not become a _man-in-the-middle_ issue.

Basically, we just need to make sure all the features we support a configurable and not implicit. The only exception is trying to default to `Rails.cache`, which we can remove in the nearest following release.

There's no need for `disabled:` option, if we pass `store: false`, the serializers should not try to call `cache_store_instance.fetch()`. Let's keep the code amount to minimum please. Might have to cleanup the obsolete options though, consider `cache_store_options.except!(:enabled)`

**Comment by doits** [comment]
> Also we should drop the default values for cache options, consider the case were the provided store: doesn't support them when calling fetch()

Exactly, this was only added for backwards compatibility, which I though was an important concern.

> There's no need for disabled: option, if we pass store: false, the serializers should not try to call cache_store_instance.fetch(). Let's keep the code amount to minimum please. 

Yes, this was only a suggestion for backwards compatibility again. With the new syntax you can pass `store: false`, but there probably is old code which has `cache_store enabled: false` somewhere out in the wild. Which will break.

---

I updated it just now to simply ignore the old options which I think is what you suggested with `cache_store_options.except!(:enabled)`. But I added a warning at least that they are ignored now.

I don't think this is the best solution, because for me "deprecating" means that if old options are used, they behave exactly like before, but the user is warned it will not be so in the future. Now old options are simply ignored and some users will not notice the warning.

So in fact those options are not deprecated but actually removed, which should either really hard fail or they should be passed to `.fetch` like every other option, because that is the new behavior. Maybe a cache store accepts those options for `.fetch()`?

I don't think there is a good middle ground here where being just *a little bit* backwards compatible and ignoring some old options will *not* lead to confusion in the long term. Only effect is that somebody might not notice the changed behavior until it silently triggers a bug for him.

File: lib/fast_jsonapi/object_serializer.rb [hunk 1]
```diff
@@ -176,9 +175,23 @@ def set_id(id_name = nil, &block)
...
 
       def cache_options(cache_options)
-        self.cached = cache_options[:enabled] || false
-        self.cache_length = cache_options[:cache_length] || 5.minutes
-        self.race_condition_ttl = cache_options[:race_condition_ttl] || 5.seconds
+        cache_options = cache_options.dup
+
+        if cache_options.key?(:store)
+          self.cache_store_instance = cache_options.delete(:store)
+        elsif !cache_store_instance
```
Comment (stas) [inline lib/fast_jsonapi/object_serializer.rb]: @doits why do we need an `elsif` here? Looks like `else` is just enough...

File: lib/fast_jsonapi/object_serializer.rb [hunk 2]
```diff
@@ -176,9 +175,23 @@ def set_id(id_name = nil, &block)
...
+          warn('DEPRECATION WARNING: `store:` is required, we will default to `Rails.cache`.')
+          self.cache_store_instance =  Rails.cache
+        end
+
+        %i[enabled cache_length].each do |key|
+          if cache_options.key?(key)
```
Comment (stas) [inline lib/fast_jsonapi/object_serializer.rb]: Consider `warn(...) unless cache_options.delete(key).nil?`, this way there's no need to have a nested block and we get rid of the extra call.

File: lib/fast_jsonapi/object_serializer.rb [hunk 3]
```diff
@@ -176,9 +175,23 @@ def set_id(id_name = nil, &block)
...
+            warn("DEPRECATION WARNING: `#{key}:` is a deprecated cache option and has no effect anymore.")
+            cache_options.delete(key)
+          end
+        end
+
+        self.cache_store_options = (cache_store_options || {}).merge(cache_options)
```
Comment (stas) [inline lib/fast_jsonapi/object_serializer.rb]: I'm not sure we need this and the initial `cache_options = cache_options.dup`. Plus merging the configuration every time might cause mutation problems on inherited serializer classes. Consider the case where one serializer inherits the other one but it has different cache options, calling `merge()` will carry over anything that was not overwritten.

Let's just keep `self.cache_store_options = cache_options.dup` to avoid any unexpected behavior for now.

File: lib/fast_jsonapi/object_serializer.rb (hunk 1 above)
Comment (doits) [reply lib/fast_jsonapi/object_serializer.rb]: It is there in case of inheritance. There might be a store already from parent, so we do not have to warn if users do this:

```rb
class BaseSerializer
  include FastJsonApi::ObjectSerializer

  cache_options store: MyStore.new, expires_in: 2.hours
end

class UserSerializer < BaseSerializer
  cache_options expires_in: 5.minutes
end
```

File: lib/fast_jsonapi/object_serializer.rb (hunk 2 above)
Comment (doits) [reply lib/fast_jsonapi/object_serializer.rb]: Yeah, I had this on my mind, too, but then the `.delete` is on the right side and therefore a little bit hidden ... I can change it though, makes the code shorter.

File: lib/fast_jsonapi/object_serializer.rb (hunk 3 above)
Comment (doits) [reply lib/fast_jsonapi/object_serializer.rb]: The initial dup might not be necessary but it doesn't hurt either. It only makes sure the hash is not modified ... who knows whether users pass some static configuration hash as an option, then we shouldn't delete keys from it. It is just some safety measure.

> Consider the case where one serializer inherits the other one but it has different cache options, calling merge() will carry over anything that was not overwritten.

I'd say exactly this is how it should be, or not? It inherits every option but can change it, otherwise what is inherited goes. This is how it was before, too.

Otherwise I must always pass a `store:`, too, if I call `cache_options`. Which makes me understand your comment above better about the `elsif`.

So you think you should have to pass *all* options on every call to `cache_options`, even on inheritance?

**Comment by doits** [comment]
> We will need to update the docs on the way cache_options(...) works now though... Either in this or in a separate PR.

It should be done, here. Some specs are missing, too, so we are not there yet. I will add them when the implementation is ready.

File: lib/fast_jsonapi/object_serializer.rb (hunk 3 above)
Comment (stas) [reply lib/fast_jsonapi/object_serializer.rb]: @doits yep, I agree, let's be explicit, or we risk getting in trouble, let's reset the `cache_store_options` on every call to avoid surprises.

File: lib/fast_jsonapi/object_serializer.rb (hunk 1 above)
Comment (doits) [reply lib/fast_jsonapi/object_serializer.rb]: As per https://github.com/fast-jsonapi/fast_jsonapi/pull/52#discussion_r372516561 this is not possible anymore.

File: lib/fast_jsonapi/object_serializer.rb (hunk 2 above)
Comment (doits) [reply lib/fast_jsonapi/object_serializer.rb]: Found a way to remove the `.delete` calls, so this is clearer now.

File: lib/fast_jsonapi/object_serializer.rb (hunk 3 above)
Comment (doits) [reply lib/fast_jsonapi/object_serializer.rb]: I don't think it is a good design choice from the point of inheritance, but I want to have it finished, too, so I implemented it.

File: spec/shared/contexts/movie_context.rb [hunk 4]
```diff
@@ -216,7 +216,7 @@ class CachingMovieSerializer
       belongs_to :owner, record_type: :user
       belongs_to :movie_type
 
-      cache_options enabled: true
+      cache_options store: ActiveSupport::Cache::MemoryStore.new
```
Comment (stas) [inline spec/shared/contexts/movie_context.rb]: @doits let's add some extra options here, ex: `expires_in: 5`, I think the tests will fail based on the
https://github.com/fast-jsonapi/fast_jsonapi/pull/52/files#diff-0d22ea72f1e72c75e7887cdbcba46871R70

I think there's a missing __*__ in `record_hash = cache_store_instance.fetch(record, *cache_store_options)`

Comment (doits) [reply spec/shared/contexts/movie_context.rb]: Yeah, tests for all the new functionality are missing yet (and might reveal some more errors). I'll add tests if the implementation is ready.

Comment (stas) [reply spec/shared/contexts/movie_context.rb]: Sorry, let me know when I could take a final look :see_no_evil: 

Thanks again for bearing with me :heart:

**Comment by stas** [comment]
@doits I'd like to get this merged soon. Please let me know if you're ok me picking up the work you left. Ty :bowing_man:

**Comment by doits** [comment]
@stas I had a week off and try to work on it this week, adding specs at least.

I still was not satisfied with breaking backwards compatibility though. While thinking about it I had the following idea to make everything a little bit simpler and 100 % backwards compatible: Now that we only have `cache_store_instance` and `cache_store_options` without any logic behind it, maybe having a new setter for them would be easier?

```rb
class UserSerializer
  include FastJsonApi::ObjectSerializer

  def cache_store(store, options = {})
    self.cache_store_instance = store
    self.cache_store_options = options
  end

  # enable caching
  cache_store ActiveSupport::Cache::MemoryStore.new, { namespace: 'fastjson' }
end
```

This would make the old `cache_options` deprecated much easier with 100 % backwards compatibility:

```rb
      def cache_options(cache_options)
        warn('DEPRECATION WARNING: `cache_options` is deprecated and will be removed soon, please use `cache_store`. See <link to docs>')

        self.cache_store_instance = cache_options[:enabled] ? Rails.cache : nil
        self.cache_store_options = {
          expires_in: cache_options[:cache_length] || 5.minutes
          race_condition_ttl: cache_options[:race_condition_ttl] || 5.seconds
        }
      end
```

With that we get:

- 100 % backwards compatibility, e.g. setting the same old defaults or stuff like `cache_options enabled: Rails.env.production?` will continue to work exactly as before with `cache_options`
- Simpler deprecation logic: Old method will be removed, new method will stay
- With new naming of setter it is clear the logic behind it changed, no confusion with old docs etc.
- Every option can be passed to `.fetch` as is, we don't have to remove the old cache option keys from it

What do you think?

**Comment by stas** [comment]
@doits I'm not sure I'm sharing your concern. It's pretty normal to deprecate a param/option of a method interface. Folks have a very simple and straight-forward way to upgrade or stick to the current version.

Again, I see the current interface a result of a poor API design and we need to start moving away from it. There's literally 0 benefit in keeping things the way they are now and I did not see any argument that would change my mind (still very open to it though, but I don't want to keep this PR open forever...).

**Comment by doits** [comment]
@stats this is exactly how I think, too, though the current implementation gave a false sense of backwards compatibility which was really bothering me. Because it used `Rails.cache` as a fallback but not the other options, and especially ignored `enabled: false`. Even if this was a poor API, it should be fully backwards compatible until removed. Or removed right now, because like you say users can simply stick to the old version until they adapt to the new API.

But you are right we do not need another method for this, fallback can be implemented inside the old `cache_options` method. Which I did just now:

- If no `store:` option is passed, we assume the user is using the old API and behave exactly like it was before but with deprecation messages.
- If a `store:` is passed, we assume the user is using the new API and do not have to care about old behavior

I think this is now the best of both worlds: Simple implementation for the new API and a 100 % backwards compatible fallback.

---

If this implementation is OK for you, I will rebase all this into one meaningful commit with a proper description and update the docs.

Do you think we need specs for the old cache options behaviour / Rails.cache fallback? And do we need some more specs for new caching options?

**Comment by doits** [comment]
... I just added one spec to test that `cache_options enabled: true` still works. I made it a self sustaining block so it can be easily removed later.

**Comment by stas** [comment]
@doits are we ready here? I'd like to merge it :)

**Comment by doits** [comment]
@stas yeah I think so if I remember correctly, specs and docs are added and it should work ;-)
